pkg/containerutil/cp_linux.go

```diff
@@ -17,163 +17,200 @@
+}
+
+// CopyFiles implements `nerdctl cp`
+// It currently depends on the following assumptions:
+// - linux only
+// - tar binary exists on the system
+// - nsenter binary exists on the system
+// - if rootless, the container is running (aka: /proc/pid/root)
+func CopyFiles(ctx context.Context, client *containerd.Client, container containerd.Container, options types.ContainerCpOptions) (err error) {
+	// We do rely on the tar binary as a shortcut - could also be replaced by archive/tar, though that would mean
+	// we need to replace nsenter calls with re-exec
 	tarBinary, isGNUTar, err := tarutil.FindTarBinary()
 	if err != nil {
 		return err
 	}
+
 	log.G(ctx).Debugf("Detected tar binary %q (GNU=%v)", tarBinary, isGNUTar)
-	var srcFull, dstFull, root, mountDestination, containerPath string
-	var cleanup func()
-	task, err := container.Task(ctx, nil)
+
+	// Actually, this can happen if the container being passed has been deleted since in a racy way
+	conSpec, err := container.Spec(ctx)
 	if err != nil {
-		// FIXME: Rootless does not support copying into/out of stopped/created containers as we need to nsenter into the user namespace of the
-		// pid of the running container with --preserve-credentials to preserve uid/gid mapping and copy files into the container.
+		return errors.Join(ErrShouldNeverHappen, err)
+	}
+
+	// Try to get a running container root
+	root, pid, err := getRoot(ctx, container)
+	// If the task is "not found" (for example, if the container stopped), we will try to mount the snapshot
+	// Any other type of error from Task() is fatal here.
+	if err != nil && !errdefs.IsNotFound(err) {
+		return errors.Join(ErrShouldNeverHappen, err)
+	}
+
+	log.G(ctx).Debugf("We have root %s and pid %d", root, pid)
+
+	// If we have no root:
+	// - bail out for rootless
+	// - mount the snapshot for rootful
+	if root == "" {
+		// FIXME: Rootless does not support copying into/out of stopped/created containers as we need to nsenter into
+		// the user namespace of the pid of the running container with --preserve-credentials to preserve uid/gid
+		// mapping and copy files into the container.
 		if rootlessutil.IsRootless() {
-			return errors.New("cannot use cp with stopped containers in rootless mode")
-		}
-		// if the task is simply not found, we should try to mount the snapshot. any other type of error from Task() is fatal here.
-		if !errdefs.IsNotFound(err) {
-			return err
+			return ErrRootlessCannotCp
 		}
-		if container2host {
-			containerPath = src
-		} else {
-			containerPath = dst
-		}
-		// Check if containerPath is in a volume
-		root, mountDestination, err = getContainerMountInfo(ctx, container, containerPath, container2host)
+
+		// See similar situation above. This may happen if we are racing against container deletion
+		var conInfo containers.Container
+		conInfo, err = container.Info(ctx)
 		if err != nil {
-			return err
+			return errors.Join(ErrShouldNeverHappen, err)
```

Comment:
why is this a `ErrShouldNeverHappen` error?

Reply:
The only situation I can think of where you would get an error here is if the container that was still live at the time we called `CopyFiles` has since been deleted.
As mentioned above, while this is not impossible, it is incredibly unlikely.

You are right though that it is 0 probability, and I can wrap that in another condition instead.

Reply:
Changing these to ErrContainerVanished